src/Provider/MapQuest/MapQuest.php

```diff
@@ -56,46 +84,91 @@
+    public function __construct(HttpClient $client, string $apiKey, bool $licensed = false, bool $useRoadPosition = false)
     {
         if (empty($apiKey)) {
             throw new InvalidCredentials('No API key provided.');
         }
 
         $this->apiKey = $apiKey;
         $this->licensed = $licensed;
+        $this->useRoadPosition = $useRoadPosition;
         parent::__construct($client);
     }
 
     /**
      * {@inheritdoc}
      */
-    public function geocodeQuery(GeocodeQuery $query): Collection
+    public function geocodeQuery(CommonGeocodeQuery $query): Collection
     {
-        $address = $query->getText();
+        $params = static::DEFAULT_GEOCODE_PARAMS;
+        $params[static::KEY_API_KEY] = $this->apiKey;
+
+        $options = static::DEFAULT_GEOCODE_OPTIONS;
+        $options[static::KEY_MAX_RESULTS] = $query->getLimit();
+
+        $useGetQuery = true;
+
+        $address = $this->extractAddressFromQuery($query);
+        if ($address instanceof Location) {
+            $params[static::KEY_LOCATION] = $this->mapAddressToArray($address);
+            $options[static::KEY_INTL_MODE] = static::MODE_5BOX;
+            $useGetQuery = false;
+        } else {
+            $addressAsText = $query->getText();
 
-        // This API doesn't handle IPs
-        if (filter_var($address, FILTER_VALIDATE_IP)) {
-            throw new UnsupportedOperation('The MapQuest provider does not support IP addresses, only street addresses.');
+            if (!$addressAsText) {
+                throw new InvalidArgument('Cannot geocode empty address');
+            }
+
+            // This API doesn't handle IPs
+            if (filter_var($addressAsText, FILTER_VALIDATE_IP)) {
+                throw new UnsupportedOperation('The MapQuest provider does not support IP addresses, only street addresses.');
+            }
+
+            $params[static::KEY_LOCATION] = $addressAsText;
         }
 
-        if ($this->licensed) {
-            $url = sprintf(self::LICENSED_GEOCODE_ENDPOINT_URL, urlencode($address), $query->getLimit(), $this->apiKey);
+        $bounds = $query->getBounds();
+        if ($bounds instanceof Bounds) {
+            $options[static::KEY_BOUNDING_BOX] = $this->mapBoundsToArray($bounds);
+            $useGetQuery = false;
+        }
+
+        if ($useGetQuery) {
+            $params = $this->addOptionsForGetQuery($params, $options);
+
+            return $this->executeGetQuery(static::GEOCODE_ENDPOINT, $params);
         } else {
-            $url = sprintf(self::OPEN_GEOCODE_ENDPOINT_URL, urlencode($address), $query->getLimit(), $this->apiKey);
+            $params = $this->addOptionsForPostQuery($params, $options);
+
+            return $this->executePostQuery(static::GEOCODE_ENDPOINT, $params);
         }
+    }
 
-        return $this->executeQuery($url);
+    private function extractAddressFromQuery(CommonGeocodeQuery $query)
+    {
+        if ($query instanceof GetAddressInterface) {
```

Comment:
Sorry for my lack of understanding. Why do we need this check? 

Dont we get the address either way? The way you implemented `getAddress` is: 

```php
public function getAddress()
{
      return $this->getData(static::DATA_KEY_ADDRESS);
}
````

Reply:
This was my attempt at future proofing the code a bit.
This way for any `GeocodeQuery` object that implements the `GetAddressInterface` we just call the `getAddess` method, and we do not care where it is stored. If I could have changed the common `GeocodeQuery` class I would have left it at that.

The line
```
return $query->getData(GeocodeQuery::DATA_KEY_ADDRESS);
```
is just a fallback for if you are using a regular `GeocodeQuery` that does not implement the `GetAddressInterface`.

I think it is better to depend on interfaces instead of depending on a known constant in a specific class.

I implemented `getAddress` the way I did to have max interoparability with the common `GeocodeQuery` class, so that a programmer did not *have* to use the mapquest version of the `GeocodeQuery` class.